[Frontend][Runtime] A plugin system for the backend devices (#343)
PennyLaneAI/catalyst: doc/dev/custom_devices.rst

unique the entry point symbol. As an example, we use the identifier ``CustomDevice``:

.. code-block:: c++

    extern "C" Catalyst::Runtime::QuantumDevice*
    CustomDeviceFactory(const std::string &kwargs) {
        return new CustomDevice(kwargs);
    }

The entry point function acts as a factory method for the device class. Note that a plugin library
may also provide several factory methods in case it packages multiple devices into the same
library. However, it is important that the device identifier be unique, as best as possible, to
avoid clashes with other plugins.

Importantly, the ``<DeviceIdentifier>`` string in the entry point function needs to match
exactly what is supplied to the ``__quantum__rt__device("rtd_name", "<DeviceIdentifier>")``
runtime instruction in compiled user programs, or what is returned from the ``get_c_interface``
function when integrating the device into a PennyLane plugin. Please see the "Integration with
Python devices" section further down for details.

``CustomDevice(kwargs)`` serves as a constructor for your custom device, with ``kwargs``
as a string of device specifications and options, represented in Python dictionary format.

> This may contain information such as device type and a number of shots: ``{'shots': 1000}``.

Reviewer: Updated version is much clearer, I think we just need to remove `device type` from since there is no context for what it could mean:
```suggestion
An example could bet the number of shots or circuit executions, encoded as follow the following string: ``"{'shots': 1000}"``.
```
(not sure about the quotes or not show the entire thing is a string)

Author: Sure, I think adding quotes makes more sense as this highlights the fact that this is passed to device as a string.

Author: Thanks! Addressed [6db724c](https://github.com/PennyLaneAI/catalyst/pull/343/commits/6db724cdca8959e77ba934db8aec6e21c5c1f055) and [ab5c257](https://github.com/PennyLaneAI/catalyst/pull/343/commits/ab5c25771f948f00f3ecd0716188fab57553ec3b) with some other fixes to the guideline.

                                        const std::vector<QubitIdType> &, bool) override
            {
            }

            auto Expval(ObsIdType) -> double override { return 0.0; }
            auto Var(ObsIdType) -> double override { return 0.0; }
            void State(DataView<std::complex<double>, 1> &) override {}
            void Probs(DataView<double, 1> &) override {}
            void PartialProbs(DataView<double, 1> &, const std::vector<QubitIdType> &) override {}
            void Sample(DataView<double, 2> &, size_t) override {}
            void PartialSample(DataView<double, 2> &, const std::vector<QubitIdType> &, size_t) override {}
            void Counts(DataView<double, 1> &, DataView<int64_t, 1> &, size_t) override {}

            void PartialCounts(DataView<double, 1> &, DataView<int64_t, 1> &,
                                    const std::vector<QubitIdType> &, size_t) override
            {
            }

            void Gradient(std::vector<DataView<double, 1>> &, const std::vector<size_t> &) override {}
        };

In addition to implementing the ``QuantumDevice`` class, one must implement the following method

> with the name ``CustomDevice`` + ``Factory``:

Author: I like `<DeviceIdentifier>` way of describing the API :+1: 
Updated this in the entire guideline [e425745](https://github.com/PennyLaneAI/catalyst/pull/343/commits/e42574552ea513430058dec8fee1bb6392071720)

.. code-block:: c++

    extern "C" Catalyst::Runtime::QuantumDevice*
    CustomDeviceFactory(bool status, const std::string &kwargs)
    {
        return new CustomDevice(status, kwargs); }

> The use of the term ``Factory`` in ``CustomDeviceFactory`` is a runtime convention that combines the name
> of the device class, in this case, ``CustomDevice``, with the suffix ``Factory``. This signifies that the
> primary purpose of the method is to act as a factory for creating instances of the ``CustomDevice`` class.

Reviewer: ```suggestion
The entry point function acts as a factory method for the device class. Note that a plugin library
may also provide several factory methods in case it packages multiple devices into the same
library. However, it is important that the device identifier be unique, as best as possible, to
avoid clashes with other plugins. 
```

Author: Updated with some changes [e425745](https://github.com/PennyLaneAI/catalyst/pull/343/commits/e42574552ea513430058dec8fee1bb6392071720)

> .. warning::
> 
>     The runtime exclusively supports factory methods with the name of the device class followed by ``Factory``.
>     Please ensure that your method adheres to this naming convention.

Reviewer: ```suggestion
Importantly, the ``<DeviceIdentifier>`` string in the entry point function needs to match
exactly what is supplied to the ``__quantum__rt__device("rtd_name", "<DeviceIdentifier>")``
runtime instruction in compiled user programs, or what is returned from the ``get_c_interface``
function when integrating the device into a PennyLane plugin (see section further down).
```

Reviewer 2: @dime10 curious what your reason is here for removing the `.. warning::` admonition? No strong feelings either way, but I often find that admonitions, if not overused, allow important information to 'stand out' better. Especially in reference documentation, that is often skim read

Reviewer: Oh I didn't mean for the change to specifically remove it, I just rewrote the instructions into flowing text as follows:

> In addition to implementing the ``QuantumDevice`` class, one must implement an entry point for the
> device library with the name ``<DeviceIdentifier>Factory``, where ``<DeviceIdentifier>`` is used to
> unique the entry point symbol. As an example, we use the identifier ``CustomDevice``:
>
> ```py
> extern "C" Catalyst::Runtime::QuantumDevice*
> CustomDeviceFactory(bool status, const std::string &kwargs) {
>    return new CustomDevice(status, kwargs);
> }
> ```
>
> The entry point function acts as a factory method for the device class. Note that a plugin library
> may also provide several factory methods in case it packages multiple devices into the same
> library. However, it is important that the device identifier be unique, as best as possible, to
> avoid clashes with other plugins. 
>
> Importantly, the ``<DeviceIdentifier>`` string in the entry point function needs to match
> exactly what is supplied to the ``__quantum__rt__device("rtd_name", "<DeviceIdentifier>")``
> runtime instruction in compiled user programs, or what is returned from the ``get_c_interface``
> function when integrating the device into a PennyLane plugin (see section further down).
>
> Additionally, the entry point function must accept two arguments defined as below, which
> can be forwarded to the device constructor (e.g. ``CustomDevice(status, kwargs)``) if
> appropriate:
> - ``status``: A boolean to control the initial value of the tape recording system in the runtime.
>  If ``true``, this will cache the entire operations and observables in a program.
>  This may be used to compute gradient of a quantum circuit with taking advantage of gradient
>  methods provided by backend devices if supported.
> - ``kwargs``: A string of device specifications and options, represented in Python dictionary format.
>  This may contain information such as device type and a number of shots:
>  ``{'device_type': 'lightning.qubit', 'shots': 0}``.

All of the parts are equally important I think if you want to implement a custom device, but let me know if you think a warning box would still make sense somewhere :)

Reviewer 2: all good! Was just curious if there was an dislike-of-admonitions reason for it lol

> ``CustomDevice(status, kwargs)`` serves as a constructor for your custom device, with

Reviewer: ```suggestion
Additionally, the entry point function must accept two arguments defined as below, which
can be forwarded to the device constructor (e.g. ``CustomDevice(status, kwargs)``) if
appropriate:
```

- ``status``: A boolean to control the initial value of the tape recording system in the runtime.
  If ``true``, this will cache the entire operations and observables in a program.
  This may be used to compute gradient of a quantum circuit with taking advantage of gradient

>   methods provided by backend devices if supported.

Reviewer: You mentioned that the tape recording system can be used by any device, but here it looks like the device still needs to keep track of the recording state?

Reviewer: I'm wondering if we need to either explain this component more, since the plugin developer is required to accept this parameter, or if we'd like to remove it somehow?

Author: Giving it some thoughts and this is not required to be part of the constructor anymore as the `StartTapeRecording` and `StopTapeRecording` methods in the QuantumDevice interface properly satisfies the  objectives of this `status` arg in the constructor. I removed this and updated the docs.

> - ``kwargs``: A string that represents a Python dictionary of device specifications in string format.
>   This may contain information about the device type and a number of shots:
>   ``{'device_type': 'lightning.qubit', 'shots': 0}``.

Reviewer: ```suggestion
- ``kwargs``: A string of device specifications and options, represented in Python dictionary format.
  This may contain information such as device type and a number of shots:
  ``{'device_type': 'lightning.qubit', 'shots': 0}``.
```

Reviewer: What is the purpose of `device_type` here?

Reviewer 3: I would like to preserve `device_type`. I at least like it for the toml files. Right now the names are `lightning.qubit.toml` and that is f`{device_type}.toml`.

Reviewer: If `device_type` is a holdover from before, or specific to only one device, we might want to use a more general example of a parameter a device might accept. But the important part is that this is not a specification, and is just showing that devices can receive arbitrary information from the program in string format.

Author: This is technically needed for `openqasm` devices to declare the type of the device (remote or local). I re-scoped this to be only appeared for `openqasm` devices (and not for lightning simulators).

> These parameters are automatically initialized in the frontend using the device specifications specified
> by users via :func:`qml.device() <pennylane.device>`.

Reviewer: ```suggestion
Note that these parameters are automatically initialized in the frontend if the library is provided as a
PennyLane plugin device (see :func:`qml.device() <pennylane.device>`).
```

Author: Updated [e425745](https://github.com/PennyLaneAI/catalyst/pull/343/commits/e42574552ea513430058dec8fee1bb6392071720)

        #include <QuantumDevice.hpp>

>         struct CustomDevice final : public Catalyst::Runtime::QuantumDevice {

Reviewer: Thanks, these are fair arguments 👍

            void Sample(DataView<double, 2> &, size_t) override {}
            void PartialSample(DataView<double, 2> &, const std::vector<QubitIdType> &, size_t) override {}
            void Counts(DataView<double, 1> &, DataView<int64_t, 1> &, size_t) override {}

            void PartialCounts(DataView<double, 1> &, DataView<int64_t, 1> &,
                                    const std::vector<QubitIdType> &, size_t) override
            {
            }

            void Gradient(std::vector<DataView<double, 1>> &, const std::vector<size_t> &) override {}
        };

In addition to implementing the ``QuantumDevice`` class, one must implement the following method:

.. code-block:: c++

    extern "C" Catalyst::Runtime::QuantumDevice*
    CustomDeviceFactory(bool status, const std::string &kwargs)
    { 
        return new CustomDevice(status, kwargs); }

where ``CustomDevice(status, kwargs)`` serves as a constructor for your custom device,
with `status` representing the initial state of the tape recording system, and `kwargs`
representing a dictionary of device specifications in string format. The destructor of

> ``CustomDevice`` will be automatically called by the runtime.

Reviewer: Btw, is the tape something that any device can use? Or is it somehow interconnect with Lighting specifically?

> ``CustomDevice`` will be automatically called by the runtime.

Author: Good points! let me update the guideline.

> ``CustomDevice`` will be automatically called by the runtime.

Author: Fixed [266cf33](https://github.com/PennyLaneAI/catalyst/pull/343/commits/266cf33edf1975967fd9a2a42ee28a5647c61300). @dime10 lmk what you think and if you're happy with the state of this PR.

> .. note::

Reviewer 2: ```suggestion
.. warning::
```

        @staticmethod
        def get_c_interface():

>             """ The tuple of davice name and the location
>                 to shared object with C/C++ implementation.
>             """
> 
>             return "CustomDevice", "full/path/to/libdummy_device.so"

Reviewer 2: ```suggestion
            """Returns a tuple consisting of the device name, and
            the location to the shared object with the C/C++ device implementation.
            """

            return "CustomDevice", "full/path/to/libdummy_device.so"
```

Reviewer 2: @maliasadi it's not entirely clear to me if the path here is relative to some directory?

Reviewer: Should be absolute. Maybe "absolute/path/to/libdummy_device.so" would be clearer than "full/path/to/libdummy_device.so"?

Reviewer 2: I think mentioning this in the text might be helpful?

Author: yep, agree! :100:

> 

Reviewer: ```suggestion

.. note::

    The first result of ``get_c_interface`` needs to to match the ``<DeviceIdentifier>``
    as described in the first section.

```

Author: Fixed along with other updates [e425745](https://github.com/PennyLaneAI/catalyst/pull/343/commits/e42574552ea513430058dec8fee1bb6392071720)
